pylint/checkers/utils.py

```diff
@@ -1805,7 +1805,14 @@ def is_typing_guard(node: nodes.If) -> bool:
 
 def is_node_in_typing_guarded_import_block(node: nodes.NodeNG) -> bool:
     """Return True if node is part for guarded `typing.TYPE_CHECKING` if block."""
-    return isinstance(node.parent, nodes.If) and is_typing_guard(node.parent)
+    ancestor = get_node_first_ancestor_of_type(node, nodes.If)
+    if ancestor is None:
+        return False
+    if not is_typing_guard(ancestor) or any(
+        node in n.nodes_of_class(nodes.NodeNG) for n in ancestor.orelse
+    ):
+        return is_node_in_typing_guarded_import_block(ancestor)
+    return True
```

Comment:
I'd like to(*) remove this method in favor of `in_type_checking_block`, which is more robust. Is the implementation of this PR any easier if it's changed to just use `in_type_checking_block`?

* PR coming shortly

Reply:
`in_type_checking_block` returns True when a name is defined any TYPE_CHECKING conditional branch. Whereas this implementation of `is_node_in_typing_guarded_import_block` is a stricter check that returns True only if a name is defined in the TYPE_CHECKING block.

In the following snippet, `is_node_in_typing_guarded_import_block` returns False while `in_type_checking_block`  returns True.
```
if TYPE_CHECKING:
    pass
else:
    import math
```

Reply:
Ah, I see. That seems like a useful improvement, but it also feels different enough from the reported issue and the variables checker changes that it should be its own PR.

My original concern is that we shouldn't have a mixture of utilities doing very similar things. (For instance, only `in_type_checking_block` is robust against importing `typing` under an alias, like `if t.TYPE_CHECKING`, which at least one of the packages we regularly test in the primer against does.)

Would something like this work?

```diff
diff --git a/pylint/checkers/utils.py b/pylint/checkers/utils.py
index b7691f88e..1d3f13d20 100644
--- a/pylint/checkers/utils.py
+++ b/pylint/checkers/utils.py
@@ -1931,6 +1931,8 @@ def in_type_checking_block(node: nodes.NodeNG) -> bool:
     for ancestor in node.node_ancestors():
         if not isinstance(ancestor, nodes.If):
             continue
+        if node in ancestor.orelse:
+            return in_type_checking_block(ancestor)
         if isinstance(ancestor.test, nodes.Name):
             if ancestor.test.name != "TYPE_CHECKING":
                 continue
```

***

Finally, if we go down this path, it's worth noting that we're not distinguishing negations like this:

```python
if not TYPE_CHECKING:
    pass
else:
    import math
```

Reply:
That was my concern as well, I think this PR is overloaded. 😅 Shall I remove the changes that is meant to address the other false negative case and focus on the VariableChecker changes first?

Reply:
Separating would be very helpful, thank you! You could open a separate issue for your Case 2 in the body of your PR.

Reply:
Gotcha, apologies for the confusion!